### ISSUE TITLE
P1: RealVis IP-Adapter — blend images in CLIP embedding space with normalized weights

### ISSUE THREAD

**Issue by ford442** [ford442/clip_stacker#180]
## Priority
**P1**

## Type
**Feature / Correctness**

## Complexity
**M** (3–5 days including qualitative evaluation)

## Problem

Multi-image combining in `RealVis_v5.0_BF16_IP/ip_adapter/ip_adapter.py:273` is one line:

```python
image_prompt_embeds = torch.cat(image_prompt_embeds_list).mean(dim=0).unsqueeze(0)
```

Four distinct problems, roughly in order of impact.

### 1. Averaging happens in the wrong space

`ImageProjModel.forward` ends in a `LayerNorm` (`ip_adapter.py:21`). The code averages N **post-LayerNorm** token sets. The mean of N normalized vectors has a smaller norm than any individual one — roughly `1/√N` for uncorrelated inputs. So combining 5 images produces a **weaker** conditioning signal than using 1, which is the washed-out blend users see today.

The adapter was trained on projections of single CLIP image embeddings. Feeding it a shrunken average is out-of-distribution.

**Fix:** average the raw `clip_image_embeds` **before** `image_proj_model`, then project once. Also N× cheaper — one projection instead of N.

### 2. Per-image scales are not normalized

`scale_i` multiplies each embedding (`ip_adapter.py:244`, `:251`, …) and then everything is divided by N. Effective weight is `scale_i / N`, so moving one image's slider changes **total** IP strength as a side effect of changing the balance between images. The two knobs are not separable.

**Fix:** `Σ(wᵢ·eᵢ) / Σwᵢ`. Global strength stays on `set_scale()` (`ip_adapter.py:118`), which is where the adapter expects it — applied in the attention processor, not baked into the embeddings.

### 3. Conditional is scaled, unconditional is not

`image_prompt_embeds_i` is multiplied by `scale_i`; `uncond_image_prompt_embeds_i` never is (`ip_adapter.py:280`). Under classifier-free guidance the two branches then have mismatched magnitude, distorting `uncond + g·(cond − uncond)` in a way that is not the "strength" control users think they are turning.

Related waste: every `uncond_image_prompt_embeds_i` is `image_proj_model(torch.zeros_like(...))` (`ip_adapter.py:115`) — **identical by construction** for a given shape. Averaging N copies of the same tensor is a no-op costing N−1 forward passes.

### 4. Averaging destroys per-image identity

Even done correctly, a mean answers "blend these moods," not "combine these subjects." `IPAttnProcessor.__call__` already splits conditioning by `self.num_tokens` (`attention_processor.py:~137`):

```python
end_pos = encoder_hidden_states.shape[1] - self.num_tokens
encoder_hidden_states, ip_hidden_states = encoder_hidden_states[:, :end_pos, :], encoder_hidden_states[:, end_pos:, :]
```

Setting `num_tokens = 4 * N` on every processor at generate time and **concatenating** token sets lets cross-attention select per-image features spatially instead of collapsing them. This is the real fix for multi-subject composition.

## Proposed solution

**Phase A — correctness (fixes 1–3).** Weighted mean of pre-projection CLIP embeddings, normalized by `Σw`; single `image_proj_model` call; compute `uncond` once; leave global strength to `set_scale`.

**Phase B — token concat (fixes 4).** Add a `combine_mode` parameter: `'mean'` (Phase A) or `'concat'`. In concat mode, set `num_tokens` on all `IPAttnProcessor` instances before the call and apply per-image weights as per-token scaling. **Keep both** — they produce genuinely different, both-useful results. Expose as a radio in the Gradio UI.

Guard the token budget: 5 images × 4 tokens is fine, but the mode should degrade gracefully rather than silently blow up attention cost.

Also worth fixing while in this file: `if pil_image_2 != None` (`ip_adapter.py:248` and following) invokes `PIL.Image.__eq__`; should be `is not None`.

## Acceptance criteria

- [ ] Blending happens pre-projection; exactly one `image_proj_model` call for the conditional branch
- [ ] Per-image weights normalized by their sum — changing the balance does not change total IP strength
- [ ] Cond/uncond magnitudes consistent; `uncond` computed once
- [ ] Single-image generation is unchanged or better (regression check against current output at a fixed seed)
- [ ] Multi-image results demonstrably stronger than the current mean at equal settings — side-by-side grids at 2/3/5 images in the PR
- [ ] `combine_mode` selectable from the UI, defaulting to the Phase A mean
- [ ] Concat mode restores `num_tokens` afterward so a later mean-mode call is unaffected
- [ ] `is not None` identity checks

## New libraries / dependencies

None.

## Compounds with

Pairs with the RealVis UI-wiring issue filed alongside this one — several generation controls in that Space are currently inert, which makes evaluating any change to blending harder than it should be.
